### PR TITLE
Use variables in Kibana index provisioning

### LINKED ISSUE #224: Variabilize api hostname and index id from Kibana index job
Acutally `INDEX_ID` is hardcoded, and the api hostname for Kibana is fetched from environnment variables which is not the best for futur maintenance.

### PR DESCRIPTION
**Pull request by alxf**
Fix #224

### CODE REVIEW

File: roles/cluster_logging/files/kibana/provision-index.sh [hunk 1]
```diff
@@ -6,9 +6,11 @@
 
 set -ue
 
-INDEX_ID="logstash-*"
+if [ ! -z "$INDEX_PATTERN" ]; then
```
Comment (NicolasT) [inline roles/cluster_logging/files/kibana/provision-index.sh]: Pretty sure this doesn't work with `set -u` :smiley: 

Also, no real reason for this... `logstash-*` is not some 'magically correct' value. Rather have the whole script fail when a required 'config value' is not set.

File: roles/cluster_logging/templates/kibana-index-provisioning-configmap.yml.j2 [hunk 2]
```diff
@@ -7,5 +7,7 @@ metadata:
     heritage: MetalK8s
     component: index-provisioning
 data:
+  index.pattern: {{ kibana_index_pattern }}
```
Comment (NicolasT) [inline roles/cluster_logging/templates/kibana-index-provisioning-configmap.yml.j2]: Not sure it makes sense setting things through the `ConfigMap`. Just set them in `env` in the `Job` template.

File: roles/cluster_logging/templates/kibana-index-provisioning-job.yml.j2 [hunk 3]
```diff
@@ -19,6 +19,17 @@ spec:
         - name: provision-index
           image: appropriate/curl
           command: ['sh', '/scripts/provision-index.sh']
+          env:
+           - name: INDEX_PATTERN
+             valueFrom:
```
Comment (NicolasT) [inline roles/cluster_logging/templates/kibana-index-provisioning-job.yml.j2]: See above. This seems like a useless indirection.

**Comment by NicolasT** [comment]
Retarget to `development/1.1`? It's not really a 'bugfix' for 1.0.

**Review by Zempashi — DISMISSED** [review]

File: ChangeLog.rst [hunk 4]
```diff
@@ -3,10 +3,16 @@
...
 ==============================
 
+
 Features added
 --------------
 :ghpull:`225` - Figure out the "fail fast" option of ansible (:ghissue:`129`)
 
+Bugs Fixed
+----------
+:ghissue:`224` - variabilize the kibana index pattern and service name in
```
Comment (NicolasT) [inline ChangeLog.rst]: `Kibana`

File: docs/architecture/cluster-services.rst [hunk 5]
```diff
@@ -168,8 +168,4 @@ Kibana
 To give operators access to the logs stored in ElasticSearch, a `Kibana`_
 instance is provided.
 
-.. note:: When accessing Kibana for the first time, an *index pattern* for the
```
Comment (NicolasT) [inline docs/architecture/cluster-services.rst:171]: Good call!

**Review by NicolasT — CHANGES_REQUESTED** [review]
Minor change requested, otherwise LGTM

**Review by NicolasT — APPROVED** [review]
Maybe consider a squash-merge :)

**Review by NicolasT — CHANGES_REQUESTED** [review]
I just tried to run this on my existing 1.0.0 cluster, and it fails:

```

TASK [cluster_logging : apply the Kibana job setup] *******************************************************************
ok: [metalk8s-master-01] => (item={u'md5sum': u'ada28baf671387a9477bbe7caaf66458', u'uid': 0, u'owner': u'root', 'diff': [], u'size': 2545, u'group': u'root', '_ansible_item_label': u'kibana-index-provisioning-configmap.yml', 'failed': False, u'state': u'file', u'gid': 0, u'secontext': u'system_u:object_r:etc_t:s0', u'invocation': {u'module_args': {u'directory_mode': None, u'force': True, u'remote_src': None, u'_original_basename': u'kibana-index-provisioning-configmap.yml.j2', u'owner': None, u'follow': False, u'local_follow': None, u'group': None, u'unsafe_writes': None, u'setype': None, u'content': None, u'serole': None, u'dest': u'/etc/kubernetes/addons/kibana/kibana-index-provisioning-configmap.yml', u'selevel': None, u'regexp': None, u'validate': None, u'src': u'/home/centos/.ansible/tmp/ansible-tmp-1533342757.31-192752873956431/source', u'checksum': u'd58990c56f6eb5ccdb01343f659be12d79113e22', u'seuser': None, u'delimiter': None, u'mode': None, u'attributes': None, u'backup': False}}, '_ansible_parsed': True, u'dest': u'/etc/kubernetes/addons/kibana/kibana-index-provisioning-configmap.yml', '_ansible_item_result': True, u'src': u'/home/centos/.ansible/tmp/ansible-tmp-1533342757.31-192752873956431/source', '_ansible_no_log': False, u'checksum': u'd58990c56f6eb5ccdb01343f659be12d79113e22', u'changed': True, 'item': u'kibana-index-provisioning-configmap.yml', u'mode': u'0644', '_ansible_ignore_errors': None}) => {"changed": false, "item": {"changed": true, "checksum": "d58990c56f6eb5ccdb01343f659be12d79113e22", "dest": "/etc/kubernetes/addons/kibana/kibana-index-provisioning-configmap.yml", "diff": [], "failed": false, "gid": 0, "group": "root", "invocation": {"module_args": {"_original_basename": "kibana-index-provisioning-configmap.yml.j2", "attributes": null, "backup": false, "checksum": "d58990c56f6eb5ccdb01343f659be12d79113e22", "content": null, "delimiter": null, "dest": "/etc/kubernetes/addons/kibana/kibana-index-provisioning-configmap.yml", "directory_mode": null, "follow": false, "force": true, "group": null, "local_follow": null, "mode": null, "owner": null, "regexp": null, "remote_src": null, "selevel": null, "serole": null, "setype": null, "seuser": null, "src": "/home/centos/.ansible/tmp/ansible-tmp-1533342757.31-192752873956431/source", "unsafe_writes": null, "validate": null}}, "item": "kibana-index-provisioning-configmap.yml", "md5sum": "ada28baf671387a9477bbe7caaf66458", "mode": "0644", "owner": "root", "secontext": "system_u:object_r:etc_t:s0", "size": 2545, "src": "/home/centos/.ansible/tmp/ansible-tmp-1533342757.31-192752873956431/source", "state": "file", "uid": 0}, "msg": "success: configmap \"kibana-index-provisioning\" configured"}
failed: [metalk8s-master-01] (item={u'md5sum': u'371dce745a372a5fd744abe2318146dd', u'uid': 0, u'owner': u'root', 'diff': [], u'size': 868, u'group': u'root', '_ansible_item_label': u'kibana-index-provisioning-job.yml', 'failed': False, u'state': u'file', u'gid': 0, u'secontext': u'system_u:object_r:etc_t:s0', u'invocation': {u'module_args': {u'directory_mode': None, u'force': True, u'remote_src': None, u'_original_basename': u'kibana-index-provisioning-job.yml.j2', u'owner': None, u'follow': False, u'local_follow': None, u'group': None, u'unsafe_writes': None, u'setype': None, u'content': None, u'serole': None, u'dest': u'/etc/kubernetes/addons/kibana/kibana-index-provisioning-job.yml', u'selevel': None, u'regexp': None, u'validate': None, u'src': u'/home/centos/.ansible/tmp/ansible-tmp-1533342761.62-157363397757258/source', u'checksum': u'665b2c7ca8bd817275bc9e4272e13e6fe2d99514', u'seuser': None, u'delimiter': None, u'mode': None, u'attributes': None, u'backup': False}}, '_ansible_parsed': True, u'dest': u'/etc/kubernetes/addons/kibana/kibana-index-provisioning-job.yml', '_ansible_item_result': True, u'src': u'/home/centos/.ansible/tmp/ansible-tmp-1533342761.62-157363397757258/source', '_ansible_no_log': False, u'checksum': u'665b2c7ca8bd817275bc9e4272e13e6fe2d99514', u'changed': True, 'item': u'kibana-index-provisioning-job.yml', u'mode': u'0644', '_ansible_ignore_errors': None}) => {"changed": false, "item": {"changed": true, "checksum": "665b2c7ca8bd817275bc9e4272e13e6fe2d99514", "dest": "/etc/kubernetes/addons/kibana/kibana-index-provisioning-job.yml", "diff": [], "failed": false, "gid": 0, "group": "root", "invocation": {"module_args": {"_original_basename": "kibana-index-provisioning-job.yml.j2", "attributes": null, "backup": false, "checksum": "665b2c7ca8bd817275bc9e4272e13e6fe2d99514", "content": null, "delimiter": null, "dest": "/etc/kubernetes/addons/kibana/kibana-index-provisioning-job.yml", "directory_mode": null, "follow": false, "force": true, "group": null, "local_follow": null, "mode": null, "owner": null, "regexp": null, "remote_src": null, "selevel": null, "serole": null, "setype": null, "seuser": null, "src": "/home/centos/.ansible/tmp/ansible-tmp-1533342761.62-157363397757258/source", "unsafe_writes": null, "validate": null}}, "item": "kibana-index-provisioning-job.yml", "md5sum": "371dce745a372a5fd744abe2318146dd", "mode": "0644", "owner": "root", "secontext": "system_u:object_r:etc_t:s0", "size": 868, "src": "/home/centos/.ansible/tmp/ansible-tmp-1533342761.62-157363397757258/source", "state": "file", "uid": 0}, "msg": "error running kubectl (/usr/local/bin/kubectl --namespace=kube-ops apply --force --filename=/etc/kubernetes/addons/kibana/kibana-index-provisioning-job.yml) command (rc=1), out='', err='The Job \"kibana-index-provisioning\" is invalid: spec.template: Invalid value: core.PodTemplateSpec{ObjectMeta:v1.ObjectMeta{Name:\"kibana-index-provisioning\", GenerateName:\"\", Namespace:\"\", SelfLink:\"\", UID:\"\", ResourceVersion:\"\", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{\"app\":\"kibana\", \"component\":\"index-provisioning\", \"controller-uid\":\"7bd58465-94c2-11e8-b29e-fa163e143620\", \"heritage\":\"MetalK8s\", \"job-name\":\"kibana-index-provisioning\"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:\"\"}, Spec:core.PodSpec{Volumes:[]core.Volume{core.Volume{Name:\"scripts\", VolumeSource:core.VolumeSource{HostPath:(*core.HostPathVolumeSource)(nil), EmptyDir:(*core.EmptyDirVolumeSource)(nil), GCEPersistentDisk:(*core.GCEPersistentDiskVolumeSource)(nil), AWSElasticBlockStore:(*core.AWSElasticBlockStoreVolumeSource)(nil), GitRepo:(*core.GitRepoVolumeSource)(nil), Secret:(*core.SecretVolumeSource)(nil), NFS:(*core.NFSVolumeSource)(nil), ISCSI:(*core.ISCSIVolumeSource)(nil), Glusterfs:(*core.GlusterfsVolumeSource)(nil), PersistentVolumeClaim:(*core.PersistentVolumeClaimVolumeSource)(nil), RBD:(*core.RBDVolumeSource)(nil), Quobyte:(*core.QuobyteVolumeSource)(nil), FlexVolume:(*core.FlexVolumeSource)(nil), Cinder:(*core.CinderVolumeSource)(nil), CephFS:(*core.CephFSVolumeSource)(nil), Flocker:(*core.FlockerVolumeSource)(nil), DownwardAPI:(*core.DownwardAPIVolumeSource)(nil), FC:(*core.FCVolumeSource)(nil), AzureFile:(*core.AzureFileVolumeSource)(nil), ConfigMap:(*core.ConfigMapVolumeSource)(0xc425caf980), VsphereVolume:(*core.VsphereVirtualDiskVolumeSource)(nil), AzureDisk:(*core.AzureDiskVolumeSource)(nil), PhotonPersistentDisk:(*core.PhotonPersistentDiskVolumeSource)(nil), Projected:(*core.ProjectedVolumeSource)(nil), PortworxVolume:(*core.PortworxVolumeSource)(nil), ScaleIO:(*core.ScaleIOVolumeSource)(nil), StorageOS:(*core.StorageOSVolumeSource)(nil)}}}, InitContainers:[]core.Container(nil), Containers:[]core.Container{core.Container{Name:\"provision-index\", Image:\"appropriate/curl\", Command:[]string{\"sh\", \"/scripts/provision-index.sh\"}, Args:[]string(nil), WorkingDir:\"\", Ports:[]core.ContainerPort(nil), EnvFrom:[]core.EnvFromSource(nil), Env:[]core.EnvVar{core.EnvVar{Name:\"INDEX_PATTERN\", Value:\"logstash-*\", ValueFrom:(*core.EnvVarSource)(nil)}, core.EnvVar{Name:\"KIBANA_SVC\", Value:\"KIBANA\", ValueFrom:(*core.EnvVarSource)(nil)}}, Resources:core.ResourceRequirements{Limits:core.ResourceList(nil), Requests:core.ResourceList(nil)}, VolumeMounts:[]core.VolumeMount{core.VolumeMount{Name:\"scripts\", ReadOnly:true, MountPath:\"/scripts\", SubPath:\"\", MountPropagation:(*core.MountPropagationMode)(nil)}}, VolumeDevices:[]core.VolumeDevice(nil), LivenessProbe:(*core.Probe)(nil), ReadinessProbe:(*core.Probe)(nil), Lifecycle:(*core.Lifecycle)(nil), TerminationMessagePath:\"/dev/termination-log\", TerminationMessagePolicy:\"File\", ImagePullPolicy:\"Always\", SecurityContext:(*core.SecurityContext)(nil), Stdin:false, StdinOnce:false, TTY:false}}, RestartPolicy:\"OnFailure\", TerminationGracePeriodSeconds:(*int64)(0xc42bbcb0c8), ActiveDeadlineSeconds:(*int64)(nil), DNSPolicy:\"ClusterFirst\", NodeSelector:map[string]string(nil), ServiceAccountName:\"\", AutomountServiceAccountToken:(*bool)(nil), NodeName:\"\", SecurityContext:(*core.PodSecurityContext)(0xc433c80370), ImagePullSecrets:[]core.LocalObjectReference(nil), Hostname:\"\", Subdomain:\"\", Affinity:(*core.Affinity)(nil), SchedulerName:\"default-scheduler\", Tolerations:[]core.Toleration(nil), HostAliases:[]core.HostAlias(nil), PriorityClassName:\"\", Priority:(*int32)(nil), DNSConfig:(*core.PodDNSConfig)(nil)}}: field is immutable\n'"}
```

**Comment by NicolasT** [comment]
So, https://github.com/scality/metalk8s/pull/233#pullrequestreview-143366361 actually makes sense and hence this PR is blocked for now.

The problem: we're trying to update the `env` part of the `Pod` defined in the `Job`, which is not permitted (it'd be impossible to change the environment variables of a running container!)

The playbook will need to delete the `Job` before continuing. Also, I wish the `Job` would be auto GC'ed after it completes (or some time after).

**Comment by alxf** [comment]
Hi @NicolasT, seems there is no way to cleanup job on completion, a part putting a mechanism in place to cleanup these resources periodically. So i just cleanup the job with ansible, before spawning it again. Is it a working solution for you ?

**Comment by NicolasT** [comment]
Yeah, I wish it was possible but it's not :smiley: So the solution you implemented works for me (it'd be required anyway!).

I'll try the current version on my cluster.

File: roles/cluster_logging/files/kibana/provision-index.sh [hunk 6]
```diff
@@ -4,11 +4,14 @@
 # Script to setup the default index pattern in kibana
 #
 
-set -ue
+set -e
 
-INDEX_ID="logstash-*"
+if [ ! -z "$INDEX_PATTERN" ]; then
```
Comment (NicolasT) [inline roles/cluster_logging/files/kibana/provision-index.sh]: > -z STRING: the length of STRING is zero
> -n STRING: the length of STRING is nonzero

So, it's either `if [ -z "$INDEX_PATTERN" ]` or `if [ ! -n "$INDEX_PATTERN" ]`, not this one :smiley:

**Review by NicolasT — CHANGES_REQUESTED** [review]
There's a bug... Also, will need to merge in recent changes from target branch.

**Review by NicolasT — DISMISSED** [review]

**Review by NicolasT — APPROVED** [review]

**Pull request closed without merging**